org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/DerivedResourcesUpdater.java

```diff
@@ -139,4 +157,37 @@ static void update(IProject workspaceProject, OmniEclipseProject project, Persis
         new DerivedResourcesUpdater(workspaceProject, project).update(persistentModel, monitor);
     }
 
+    /**
+     * Helper class to collect derived resources.
+     */
+    private static final class DerivedResourcesInfo {
+
+        private final IPath projectBuildDir;
+        private final Collection<IPath> nestedProjectPaths;
+        private final Collection<IPath> nestedProjectBuildDirs;
+
+        public DerivedResourcesInfo(IPath projectBuildDir, Collection<IPath> nestedProjectPaths, Collection<IPath> nestedProjectBuildDirs) {
+            this.projectBuildDir = projectBuildDir;
+            this.nestedProjectPaths = nestedProjectPaths;
+            this.nestedProjectBuildDirs = nestedProjectBuildDirs;
+        }
+
+        public IPath getProjectBuildDir() {
+            return this.projectBuildDir == null ? new Path("build") : this.projectBuildDir;
```

Comment:
If the build dir points outside to the Eclipse workspace and no linked resources references that location then `normalizeBuildDirectoryPath()` will return `null`.  Returning the `build` folder as default is not a proper solution, though. How about we make `PersistentModel.getBuildDir()` return `Optional<IPath>`?